polars/polars-io/src/parquet/read_par.rs

```diff
@@ -0,0 +1,94 @@
+use crate::mmap::{MmapBytesReader, ReaderBytes};
+use arrow::array::ArrayRef;
+use arrow::io::parquet::read;
+use arrow::io::parquet::read::{FileMetaData, MutStreamingIterator};
+use polars_core::prelude::*;
+use polars_core::utils::accumulate_dataframes_vertical;
+use polars_core::POOL;
+use rayon::prelude::*;
+use std::borrow::Cow;
+use std::convert::TryFrom;
+use std::io::Cursor;
+use std::ops::Deref;
+use std::sync::Arc;
+
+pub(crate) fn parallel_read<R: MmapBytesReader>(
+    reader: R,
+    limit: usize,
+    projection: Option<&[usize]>,
+    arrow_schema: &ArrowSchema,
+    metadata: Option<FileMetaData>,
+) -> Result<DataFrame> {
+    let reader = ReaderBytes::from(&reader);
+    let bytes = reader.deref();
+    let mut reader = Cursor::new(bytes);
+
+    let file_metadata = metadata
+        .map(Ok)
+        .unwrap_or_else(|| read::read_metadata(&mut reader))?;
+
+    let parq_fields = if let Some(projection) = projection {
+        let parq_fields = file_metadata.schema().fields();
+        Cow::Owned(
+            projection
+                .iter()
+                .map(|i| parq_fields[*i].clone())
+                .collect::<Vec<_>>(),
+        )
+    } else {
+        Cow::Borrowed(file_metadata.schema().fields())
+    };
+
+    let n_groups = file_metadata.row_groups.len();
+    let mut dfs = Vec::with_capacity(n_groups);
+
+    for row_group in 0..n_groups {
+        let columns = POOL.install(|| {
+            parq_fields
+                .par_iter()
+                .enumerate()
+                .map(|(mut field_i, field)| {
+                    if let Some(projection) = projection {
+                        field_i = projection[field_i];
+                    }
+
+                    // <IO bounded>
+
```

Comment:
@jorgecarleitao I combined the IO part with the CPU heavy part. This will likely reduce memory pressure as we have a limited amount of threads.